### DISCUSSION TITLE
Load GLB as an Object/Model to an already existing scene

### DISCUSSION THREAD

**Discussion by TiduTidu** [google/filament#3561] [Q&A] [answered]
The title states it but just to give an example on what im trying to achieve. 
Having a 'room.glb' as my base scene on the surfaceView as an environment and then adding 'stool.glb' + 'bed.glb' models in the base scene as objects. 
I also wanted to ask if its possible to use the light thats being imported with an glb file.

**Comment by romainguy** [comment, marked as answer]
You can use our gltfio library to achieve this. And yes lights will be loaded.

**Reply by TiduTidu** [reply]
Are there any code snipets you could lead me to im kinda lost. ty for the quick reply

**Reply by paulorrfilho** [reply]
I have the same question. Did you manage to achieve this? The only example loading a glTF file does not seem to be using the gltfio library mentioned here.

**Reply by romainguy** [reply]
It is using the gltfio library. It is the only way we provide to load glTF files.

But here's another sample if you want: https://github.com/romainguy/sample-materials-shop/blob/99a54fd6d30abd39fbcc54d2066a57fc8e737ed7/app/src/main/kotlin/com/curiouscreature/compose/sample/shop/MainActivity.kt#L129

In this function the app loads several glTF files.

**Reply by paulorrfilho** [reply]
I think this is the snippet you are looking for (found in the AssetLoader file). It looks like you need to create a filamentAsset -> Load that asset using ResourceLoader -> then add the filamentAsset entities to your scene. Not very well documented in terms of what are the best practices on how to use the API but that's what I got. In the example, they use the ModelViewer to indirectly load the glTF file, not clear if we still need to use the ModelViewer though.

`override fun onCreate(savedInstanceState: Bundle?) {
 
     ...
 
      assetLoader = AssetLoader(engine, MaterialProvider(engine), EntityManager.get())
 
      filamentAsset = assets.open("models/lucy.gltf").use { input ->
          val bytes = ByteArray(input.available())
          input.read(bytes)
          assetLoader.createAssetFromJson(ByteBuffer.wrap(bytes))!!
      }
 
      val resourceLoader = ResourceLoader(engine)
      for (uri in filamentAsset.resourceUris) {
          val buffer = loadResource(uri)
          resourceLoader.addResourceData(uri, buffer)
      }
      resourceLoader.loadResources(filamentAsset)
      resourceLoader.destroy()
      animator = asset.getAnimator()
      filamentAsset.releaseSourceData();
 
      scene.addEntities(filamentAsset.entities)
  }`

**Reply by paulorrfilho** [reply]
Nice! We were typing at the same time. I guess we plebs are moving in the right direction.
Thanks @romainguy

**Reply by paulorrfilho** [reply]
@romainguy the code you shared is very helpful, thanks for that.
For what I understand one would add the filamentAsset.entities to a scene and any changes to the materials after that needs to be done to the filamentAsset, correct? Meaning the scene does not hold a reference to the assets in it?

**Reply by romainguy** [reply]
You can change the materials on the entities themselves.

**Reply by paulorrfilho** [reply]
ok, but if I don't have a reference to the entity instance yet I need to find it by name through the filamentAsset, correct?

**Reply by romainguy** [reply]
You can use FilamentAsset::getName(Entity) to find the name of an entity. You can also use getFirstEntityByName and other methods to find entities.
